### ISSUE TITLE
Investigate X-Frame-Options Header Usage

### ISSUE THREAD

**Issue by rotated8** [emory-libraries/avalon#106]
Currently, Avalon sets the `X-Frame-Options` header to `ALLOWALL` by default. This is a bad practice which can allow ["clickjacking" attacks](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options). Understandably, security would like us to change this.

For us to test a change, we need to be able to show that access to content from Canvas works now, and works after the change. So, security has temporarily granted us a firewall exception which will expire on the 24th.

The next steps for this ticket are blocked until the LTI integration completes, so we can have a user test accessing content from Canvas. Then:
- [x] Add content to Avalon for the test course, and show that access from Canvas works.
~Change the `X-Frame-Options` header to `DENY` and identify any breakages to access~
Attempt to use the `Content Security Policy` header's [`frame-ancestors` directive](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors). This work includes:
- [x] Adding a ENV var for the allowed hosts for `frame-ancestors` (for now, it should be `'self' https://*.emory.edu https://emory.edu`)
- [x] Pulling the ENV var to the application config
- [x] Changing the `X-Frame-Options` header to `CSP frame-ancestors`, with the configured hosts
- [x] Checking for breakages with the Canvas LTI
Finally,
- [x] Inform security of the changes made and the effects those changes had

This is done when Security has made our firewall exception permanent.

Note that `Content Security Policy` is incompatible with Internet Explorer.

Avalon sets the `X-Frame-Options` [here](https://github.com/avalonmediasystem/avalon/blob/master/config/application.rb#L37), and unsets it for embeds [here](https://github.com/avalonmediasystem/avalon/blob/master/app/controllers/master_files_controller.rb#L95)

**Comment by rotated8** [comment]
@SolomonHD This summarizes my understanding of the issue. Is there anything else to add to this? Since we can manage the headers in the application, I do not expect this to take middleware time, unless something needs to change at the Apache level.

**Comment by rotated8** [comment]
@lovinscari This may have unintended consequences for instructors. I can cover that in standup, or we can meet to discuss it separately. Let me know which you prefer.

**Comment by SolomonHD** [comment]
From Phoung
![image.png](https://images.zenhubusercontent.com/5c191329a896da5d83d69955/f375c661-3a96-452c-8fd4-2ab0530f589e)

**Comment by libdgg** [comment]
@SolomonHD please add details to this ticket from your communications with Phuong

**Comment by rotated8** [comment]
I have updated this ticket with feedback from @lovinscari about the allowed hosts.

**Comment by SolomonHD** [comment]
There was another issue about the headers relating to the HTTP Strict-Transport-Security or HSTS. Security wanted this header set to a certain value. By default it's not set at all.

Using the F5 I am able to inject this header, it should now appear on all avalon environments. The iRule to make this happen is called HSTS and it's in the Common partition.

**Comment by libdgg** [comment]
NOTES per standup 11/19/2020
get work done and sit in a branch. once LTI work is done we can merge.
approval for this work is anyone that can login via LTI and view content (basically anyone that can accept the LTI work)

**Comment by devanshu-m** [comment]
Application code work for this is here: https://github.com/emory-libraries/avalon/commit/823c8cfc544419d912ce68900b5609d1c4a569aa

Env variable name is `CSP_FRAME_ANCESTORS`

**Comment by lovinscari** [comment]
@SolomonHD will work on environment variable work today.

**Comment by SolomonHD** [comment]
```CSP_FRAME_ANCESTORS="'self' https://*.emory.edu https://emory.edu"``` has been added to all envs. I've also added this to terraform.

**Comment by devanshu-m** [comment]
This has been deployed to test. I think the following scenario should be tested:

- [ ] login via canvas
- [ ] be able to stream a video in a course

Since headers have changed, should we test this with a fresh video upload via shib? Not sure, open to ideas.

**Comment by libdgg** [comment]
The next step for this is in ticket #54 (and in the checkbox list at the top of this ticket). Once Collin informs security and we do a final test we can close.
